File: reserved-quantity-drift/node/reserved-quantity-drift.js
/**
 * Find and repair PrestaShop reserved_quantity drift from real pending orders.
 *
 * stock_available.reserved_quantity is a running counter PrestaShop updates as a side
 * effect of order_histories inserts, not a live query. When an order state changes
 * outside the normal flow, the decrement can be skipped and the counter never comes
 * back down. This recomputes the expected reserved quantity from real open orders,
 * diffs it against the API, and repairs drift by reposting the order's own current
 * state to order_histories, which re-triggers PrestaShop's native stock recalculation.
 * Never writes reserved_quantity or physical_quantity directly. Safe to run again and again.
 *
 * Guide: https://www.allanninal.dev/prestashop/reserved-quantity-drift/
 */
import { pathToFileURL } from "node:url";

const BASE_URL = (process.env.PRESTASHOP_URL || "https://example.test").replace(/\/$/, "");
const WS_KEY = process.env.PRESTASHOP_WS_KEY || "dummy_key";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

function authHeader() {
  return "Basic " + Buffer.from(`${WS_KEY}:`).toString("base64");
}

async function apiGet(path, params) {
  const url = new URL(`${BASE_URL}/api/${path}`);
  for (const [k, v] of Object.entries(params)) url.searchParams.set(k, v);
  url.searchParams.set("output_format", "JSON");
  const res = await fetch(url, { headers: { Authorization: authHeader() } });
  if (!res.ok) throw new Error(`PrestaShop ${res.status}`);
  return res.json();
}

async function logableStateIds() {
  const data = await apiGet("order_states", { display: "full" });
  const states = data.order_states || [];
  return new Set(states.filter((s) => String(s.logable) === "1" || s.logable === true).map((s) => Number(s.id)));
}

async function openOrders(logableIds) {
  const data = await apiGet("orders", { display: "full", limit: "0,1000" });
  const orders = data.orders || [];
  return orders.filter((o) => logableIds.has(Number(o.current_state)));
}

async function orderLines(idOrder) {
  const data = await apiGet("order_details", { display: "full", "filter[id_order]": idOrder });
  return data.order_details || [];
}

async function stockRows() {
  const data = await apiGet("stock_availables", { display: "full", limit: "0,1000" });
  const rows = data.stock_availables || [];
  return rows.map((r) => ({
    id_product: Number(r.id_product),
    id_product_attribute: Number(r.id_product_attribute || 0),
    reserved_quantity: Number(r.reserved_quantity || 0),
  }));
}

/** Pure function. No I/O. See reserved-quantity-drift.test.js for fixtures. */
export function computeReservedDrift(openOrderLines, logableStateIds, stockRowsList) {
  const expected = new Map();
  for (const line of openOrderLines) {
    if (!logableStateIds.has(line.id_order_state)) continue;
    const key = `${line.id_product}:${line.id_product_attribute}`;
    let remaining = line.product_quantity - line.product_quantity_refunded;
    if (remaining < 0) remaining = 0;
    expected.set(key, (expected.get(key) || 0) + remaining);
  }

  const actualByKey = new Map();
  for (const row of stockRowsList) {
    actualByKey.set(`${row.id_product}:${row.id_product_attribute}`, row.reserved_quantity);
  }

  const keys = new Set([...expected.keys(), ...actualByKey.keys()]);
  const results = [];
  for (const key of keys) {
    const [idProduct, idProductAttribute] = key.split(":").map(Number);
    const expectedReserved = expected.get(key) || 0;
    const actualReserved = actualByKey.get(key) || 0;
    if (expectedReserved !== actualReserved) {
      results.push({
        id_product: idProduct,
        id_product_attribute: idProductAttribute,
        expected_reserved: expectedReserved,
        actual_reserved: actualReserved,
        drift: actualReserved - expectedReserved,
      });
    }
  }
  return results;
}

async function resyncOrderState(idOrder, idOrderState) {
  const url = new URL(`${BASE_URL}/api/order_histories`);
  url.searchParams.set("output_format", "JSON");
  const res = await fetch(url, {
    method: "POST",
    headers: { Authorization: authHeader(), "Content-Type": "application/json" },
    body: JSON.stringify({ order_history: { id_order: idOrder, id_order_state: idOrderState } }),
  });
  if (!res.ok) throw new Error(`PrestaShop ${res.status}`);
  return res.json();
}

async function ordersTouchingProduct(openOrdersList, idProduct, idProductAttribute) {
  const matches = [];
  for (const order of openOrdersList) {
    const lines = await orderLines(order.id);
    if (lines.some((l) => Number(l.product_id) === idProduct && Number(l.product_attribute_id || 0) === idProductAttribute)) {
      matches.push(order);
    }
  }
  return matches;
}

export async function run() {
  const ids = await logableStateIds();
  const orders = await openOrders(ids);
  const lines = [];
  for (const order of orders) {
    const idState = Number(order.current_state);
    for (const line of await orderLines(order.id)) {
      lines.push({
        id_product: Number(line.product_id),
        id_product_attribute: Number(line.product_attribute_id || 0),
        product_quantity: Number(line.product_quantity || 0),
        product_quantity_refunded: Number(line.product_quantity_refunded || 0),
        id_order_state: idState,
      });
    }
  }
  const rows = await stockRows();
  const drifted = computeReservedDrift(lines, ids, rows);

  for (const item of drifted) {
    console.warn(
      `Product ${item.id_product} attribute ${item.id_product_attribute} drift: expected=${item.expected_reserved} actual=${item.actual_reserved} (${DRY_RUN ? "would resync" : "resyncing"})`
    );
    if (!DRY_RUN) {
      const touching = await ordersTouchingProduct(orders, item.id_product, item.id_product_attribute);
      for (const order of touching) {
        await resyncOrderState(order.id, Number(order.current_state));
      }
    }
  }

  console.log(`Done. ${drifted.length} drifted product/attribute row(s) ${DRY_RUN ? "to resync" : "resynced"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}

File: reserved-quantity-drift/python/reserved_quantity_drift.py
"""Find and repair PrestaShop reserved_quantity drift from real pending orders.

stock_available.reserved_quantity is a running counter PrestaShop updates as a side
effect of order_histories inserts, not a live query. When an order state changes
outside the normal flow, the decrement can be skipped and the counter never comes
back down. This recomputes the expected reserved quantity from real open orders,
diffs it against the API, and repairs drift by reposting the order's own current
state to order_histories, which re-triggers PrestaShop's native stock recalculation.
Never writes reserved_quantity or physical_quantity directly. Safe to run again and again.

Guide: https://www.allanninal.dev/prestashop/reserved-quantity-drift/
"""
import os
import logging
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("reserved_quantity_drift")

BASE_URL = os.environ.get("PRESTASHOP_URL", "https://example.test").rstrip("/")
WS_KEY = os.environ.get("PRESTASHOP_WS_KEY", "dummy_key")
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"


def api_get(path, params):
    params = dict(params)
    params["output_format"] = "JSON"
    r = requests.get(f"{BASE_URL}/api/{path}", params=params, auth=(WS_KEY, ""), timeout=30)
    r.raise_for_status()
    return r.json()


def logable_state_ids():
    data = api_get("order_states", {"display": "full"})
    states = data.get("order_states") or []
    return {int(s["id"]) for s in states if str(s.get("logable")) in ("1", "true", "True")}


def open_orders(logable_ids):
    data = api_get("orders", {"display": "full", "limit": "0,1000"})
    orders = data.get("orders") or []
    return [o for o in orders if int(o.get("current_state", 0)) in logable_ids]


def order_lines(id_order):
    data = api_get("order_details", {"display": "full", "filter[id_order]": id_order})
    return data.get("order_details") or []


def open_order_lines(logable_ids):
    lines = []
    for order in open_orders(logable_ids):
        id_state = int(order["current_state"])
        for line in order_lines(order["id"]):
            lines.append({
                "id_product": int(line["product_id"]),
                "id_product_attribute": int(line.get("product_attribute_id") or 0),
                "product_quantity": int(line.get("product_quantity") or 0),
                "product_quantity_refunded": int(line.get("product_quantity_refunded") or 0),
                "id_order_state": id_state,
            })
    return lines


def stock_rows():
    data = api_get("stock_availables", {"display": "full", "limit": "0,1000"})
    rows = data.get("stock_availables") or []
    return [
        {
            "id_product": int(r["id_product"]),
            "id_product_attribute": int(r.get("id_product_attribute") or 0),
            "reserved_quantity": int(r.get("reserved_quantity") or 0),
        }
        for r in rows
    ]


def compute_reserved_drift(open_order_lines, logable_ids, stock_rows_list):
    """Pure function. No I/O. See test_reserved_drift.py for fixtures."""
    expected = {}
    for line in open_order_lines:
        if line["id_order_state"] not in logable_ids:
            continue
        key = (line["id_product"], line["id_product_attribute"])
        remaining = line["product_quantity"] - line["product_quantity_refunded"]
        if remaining < 0:
            remaining = 0
        expected[key] = expected.get(key, 0) + remaining

    actual_by_key = {
        (row["id_product"], row["id_product_attribute"]): row["reserved_quantity"]
        for row in stock_rows_list
    }

    keys = set(expected) | set(actual_by_key)
    results = []
    for key in keys:
        expected_reserved = expected.get(key, 0)
        actual_reserved = actual_by_key.get(key, 0)
        if expected_reserved != actual_reserved:
            id_product, id_product_attribute = key
            results.append({
                "id_product": id_product,
                "id_product_attribute": id_product_attribute,
                "expected_reserved": expected_reserved,
                "actual_reserved": actual_reserved,
                "drift": actual_reserved - expected_reserved,
            })
    return results


def resync_order_state(id_order, id_order_state):
    body = {"order_history": {"id_order": id_order, "id_order_state": id_order_state}}
    r = requests.post(
        f"{BASE_URL}/api/order_histories",
        params={"output_format": "JSON"},
        json=body,
        auth=(WS_KEY, ""),
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def orders_touching_product(open_orders_list, id_product, id_product_attribute):
    matches = []
    for order in open_orders_list:
        for line in order_lines(order["id"]):
            if int(line["product_id"]) == id_product and int(line.get("product_attribute_id") or 0) == id_product_attribute:
                matches.append(order)
                break
    return matches


def run():
    ids = logable_state_ids()
    orders = open_orders(ids)
    lines = []
    for order in orders:
        id_state = int(order["current_state"])
        for line in order_lines(order["id"]):
            lines.append({
                "id_product": int(line["product_id"]),
                "id_product_attribute": int(line.get("product_attribute_id") or 0),
                "product_quantity": int(line.get("product_quantity") or 0),
                "product_quantity_refunded": int(line.get("product_quantity_refunded") or 0),
                "id_order_state": id_state,
            })
    rows = stock_rows()
    drifted = compute_reserved_drift(lines, ids, rows)

    for item in drifted:
        log.warning(
            "Product %s attribute %s drift: expected=%s actual=%s (%s)",
            item["id_product"], item["id_product_attribute"],
            item["expected_reserved"], item["actual_reserved"],
            "would resync" if DRY_RUN else "resyncing",
        )
        if not DRY_RUN:
            touching = orders_touching_product(orders, item["id_product"], item["id_product_attribute"])
            for order in touching:
                resync_order_state(order["id"], int(order["current_state"]))

    log.info("Done. %d drifted product/attribute row(s) %s.", len(drifted), "to resync" if DRY_RUN else "resynced")


if __name__ == "__main__":
    run()
